src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/ModuleBreadcrumbs.kt

```diff
@@ -18,91 +18,178 @@
 import net.ccbluex.liquidbounce.render.utils.rainbow
-import net.ccbluex.liquidbounce.render.withColor
-import net.ccbluex.liquidbounce.utils.entity.interpolateCurrentPosition
 import net.ccbluex.liquidbounce.utils.math.toVec3
+import net.minecraft.client.render.GameRenderer
+import net.minecraft.client.render.VertexFormat.DrawMode
+import net.minecraft.client.render.VertexFormats
+import net.minecraft.client.util.math.MatrixStack
+import net.minecraft.entity.Entity
 import net.minecraft.util.math.Vec3d
 
+// TODO on tick and interpolation
 /**
  * Breadcrumbs module
  *
- * Leaves a trace behind you.
+ * Leaves traces behind players.
  */
-
 object ModuleBreadcrumbs : Module("Breadcrumbs", Category.RENDER, aliases = arrayOf("PlayerTrails")) {
 
+    private val onlyOwn by boolean("OnlyOwn", true)
     private val color by color("Color", Color4b(255, 179, 72, 255))
     private val colorRainbow by boolean("Rainbow", false)
-    private val maxLength by int("MaxLength", 500, 10..5000)
+    private val height by float("Height", 0f, 0f..2f)
+    private val alive by int("Alive", 2500, 10..10000, "ms")
+    private val fade by boolean("Fade", true)
 
-    private val positions = mutableListOf<Double>()
-    private var lastPosX = 0.0
-    private var lastPosY = 0.0
-    private var lastPosZ = 0.0
-
-    override fun enable() {
-        synchronized(positions) {
-            positions.addAll(listOf(player.x, player.eyeY, player.z))
-            positions.addAll(listOf(player.x, player.y, player.z))
-        }
-    }
+    private val positions = mutableMapOf<Entity, MutableList<TrailPart>>()
+    private val lastPositions = mutableMapOf<Entity, Array<Double>>()
 
     override fun disable() {
-        synchronized(positions) {
-            positions.clear()
-        }
+        clear()
     }
 
     val renderHandler = handler<WorldRenderEvent> { event ->
+        val time = System.currentTimeMillis()
+        update(time)
+
         val matrixStack = event.matrixStack
         val color = if (colorRainbow) rainbow() else color
 
-        synchronized(positions) {
-            renderEnvironmentForWorld(matrixStack) {
-                withColor(color) {
-                    val lines = makeLines(positions, event.partialTicks)
-                        .map { relativeToCamera(it).toVec3() }
-                        .toTypedArray()
-                    @Suppress("SpreadOperator")
-                    drawLineStrip(*lines)
-                }
+        renderEnvironmentForWorld(matrixStack) {
+            if (height > 0) {
+                RenderSystem.disableCull()
+            }
+
+            positions.values.forEach { doubles ->
+                val lines = makeAndVerifyVertexList(time, doubles)
+                    .map { Pair(relativeToCamera(it.first).toVec3(), it.second) }
+                    .toTypedArray()
+
+                @Suppress("SpreadOperator")
+                draw(matrixStack, color, *lines)
+            }
```

Comment:
This is a real performance issue since you do a render call for every entity you want to render. Each render call might take for example `5µs + vertices x Y ns/vertex`. If you have 1000 entities and do a new render call for all of them you waste `999 x 5µs = 5 ms` with just talking to the GPU.

Please do it in a single call